### ISSUE TITLE
Hyper files created using pantab cannot be published to tableau server using tableauserverclient

### ISSUE THREAD

**Issue by debola31** [innobi/pantab#71]
**Describe the bug**
When trying to publish .hyper files created using pantab using tableauserverclient you will get the generic publishing error "400011: Bad Request -- There was a problem publishing the file".

This is similar to what is described in this tableau forum thread: https://community.tableau.com/thread/324979

**To Reproduce**
1. create .hyper file with pantab
2. try publishing it using tableauserverclient

**Additional context**
The fix for this will be to do as described in the forum thread and edit the _insert_frame function in the _writer.py file to properly create the schema when defining tables.

The lines:
```
if isinstance(table, tab_api.TableName) and table.schema_name:
            connection.catalog.create_schema_if_not_exists(table.schema_name)
```
 are never executed because table.schema_name is never defined so the schema never gets created.

**Comment by debola31** [comment]
The only way I've been able to solve this issue is by changing line 114 in _writer.py to:

```
if isinstance(table, str):
        connection.catalog.create_schema('Extract')
        table = tab_api.TableName('Extract', 'Extract')
```

**Comment by WillAyd** [comment]
Can you share a minimal example in code to reproduce the issue?

Note that something similar to what you describe as fix is being done here:

https://github.com/innobi/pantab/blob/9c8771631941a683bba9203e7c4f0e50ba322e16/pantab/_writer.py#L141

Potentially could move it up but that would make things "less atomic"

**Comment by debola31** [comment]
Example:

```
example = pandas.read_csv('example.csv')
pantab.frame_to_hyper(example, "example.hyper", table="example")

tableau_auth = tableauserverclient.TableauAuth('username', 'password', 'site')
server = tableauserverclient.Server('server')
server.add_http_options({'verify': False})
    
with server.auth.sign_in(tableau_auth):
        project_id = 'project_id'
        file_path = '{}/{}'.format(str(Path.cwd()), "example.hyper")
        new_datasource = tableauserverclient.DatasourceItem(project_id)
        try:
            new_datasource = server.datasources.publish(
                    new_datasource, file_path, 'Overwrite')
        except Exception as e:
            print('Failed to publish with error: {}'.format(e))
```

It might be a problem with the hyper api but after forking you code, I found that using any other schema name other than 'Extract' would cause the .hyper files to be unable to be published to a tableau server.

e.g: 
```
if isinstance(table, str):
        connection.catalog.create_schema('Extract')
        table = tab_api.TableName('Extract', 'Extract')
```
would work but 
```
if isinstance(table, str):
        connection.catalog.create_schema('example')
        table = tab_api.TableName('example', table)
```
would not work.

Also, it may have been my data frames but table.schema_name was always null and so the if statement to create the schema never ran for me.

Thanks for the response by the way. This is a really useful library.

**Comment by WillAyd** [comment]
Can you post the traceback from the error that would occur? Wonder if the tableauserverclient is imposing that limitation here (I think historically you couldn't control the schema location with old Extract APIs)

**Comment by debola31** [comment]
For this code:

```
example = pd.read_csv('example')
pantab.frame_to_hyper(example, "example.hyper", table="example")
tableau_auth = tableauserverclient.TableauAuth('username', 'password', 'site')
server = tableauserverclient.Server('server')
server.add_http_options({'verify': False})

with server.auth.sign_in(tableau_auth):
	project_id = '8b1c9537-94b4-4f7c-9ced-d4d02b8e67f6'
	file_path = '{}/{}'.format(str(Path.cwd()), "example.hyper")
	new_datasource = tableauserverclient.DatasourceItem(project_id)
	new_datasource = server.datasources.publish(new_datasource, file_path, 'Overwrite')
```
Traceback:
```

Traceback (most recent call last):
  File "test.py", line 34, in <module>
    main()
  File "test.py", line 31, in main
    new_datasource, file_path, 'Overwrite')
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/tableauserverclient/server/endpoint/endpoint.py", line 121, in wrapper
    return func(self, *args, **kwargs)
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/tableauserverclient/server/endpoint/endpoint.py", line 159, in wrapper
    return func(self, *args, **kwargs)
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/tableauserverclient/server/endpoint/endpoint.py", line 159, in wrapper
    return func(self, *args, **kwargs)
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/tableauserverclient/server/endpoint/datasources_endpoint.py", line 206, in publish
    server_response = self.post_request(url, xml_request, content_type)
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/tableauserverclient/server/endpoint/endpoint.py", line 96, in post_request
    content_type=content_type)
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/tableauserverclient/server/endpoint/endpoint.py", line 55, in _make_request
    self._check_status(server_response)
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/tableauserverclient/server/endpoint/endpoint.py", line 70, in _check_status
    raise ServerResponseError.from_response(server_response.content, self.parent_srv.namespace)
tableauserverclient.server.endpoint.exceptions.ServerResponseError: 

        400011: Bad Request
                There was a problem publishing the file 'example.hyper'.
```

**Comment by WillAyd** [comment]
This is all great information. So I don't think any of this is actually an issue with pantab - the problem appears that the default schema name for the Hyper API to write to is "public" whereas TableauServerClient expects the schema name to be "Extract"

Note that you can verify this by doing the following:

```python
>>> pantab.frame_to_hyper(pd.DataFrame([[1]], columns=["a"]), "test.hyper", table="test")
>>> pantab.frames_from_hyper("test.hyper")
{TableName('public', 'test'):    a
0  1}
```

Instead of changing any pantab code, perhaps a more appropriate workaround is to specify the schema you want to write to:

```python
>>> from tableauhyperapi import TableName
>>> pantab.frame_to_hyper(pd.DataFrame([[1]], columns=["a"]), "test.hyper", table=TableName("Extract", "test"))
>>> pantab.frames_from_hyper("test.hyper")
{TableName('Extract', 'test'):    a
0  1}
```

Is this an issue you have opened with the TSC perhaps?

**Comment by WillAyd** [comment]
@vogelsgesang offered some insight here; looks like this is a documented requirement for publishing hyper files to Tableau Server:

https://help.tableau.com/current/api/hyper_api/en-us/docs/hyper_api_publish.html

**Comment by debola31** [comment]
Ah, good to know. I had read through the docs but somehow missed that. Thanks.

**Comment by WillAyd** [comment]
@debola31 in speaking with @vogelsgesang this restriction on the Tableau Server side will be fixed in 2020.1 so might want to give that a look
